hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/fs/viewfs/TestViewFsOverloadSchemeWithHdfsScheme.java

```diff
@@ -0,0 +1,351 @@
+    try {
+      fs.create(new Path("/nonMount/myfile"));
+      FileStatus[] ls = fs.listStatus(new Path("/nonMount"));
+      Assert.assertEquals(1, ls.length);
+      Assert.assertEquals(
+          Path.getPathWithoutSchemeAndAuthority(ls[0].getPath()).getName(),
+          "myfile");
+    } finally {
+      fs.close();
+    }
+  }
+
+  /**
+   * Create mount links as follows
+   * hdfs://localhost:xxx/HDFSUser --> hdfs://localhost:xxx/HDFSUser/
+   * hdfs://localhost:xxx/local --> file://TEST_ROOT_DIR/root/
+   *
+   * It can not find any mount link. ViewFS expects a mount point from root.
+   */
+  @Test(expected = NotInMountpointException.class, timeout = 30000)
+  public void testCreateOnRootShouldFailWhenMountLinkConfigured()
+      throws Exception {
+    final Path hdfsTargetPath = new Path(defaultFSURI + HDFS_USER_FOLDER);
+    final Path localTragetPath = new Path(localTargetDir.toURI());
+
+    createLinks(false, hdfsTargetPath, localTragetPath);
+
+    ViewFsOverloadScheme fs = (ViewFsOverloadScheme) FileSystem.get(conf);
+    try {
+      fs.create(new Path("/newFileOnRoot"));
+      Assert.fail("It should fail as root is read only in viewFS.");
+    } finally {
+      fs.close();
+    }
+  }
+
+  /**
+   * Create mount links as follows
+   * hdfs://localhost:xxx/HDFSUser --> hdfs://localhost:xxx/HDFSUser/
+   * hdfs://localhost:xxx/local --> file://TEST_ROOT_DIR/root/
+   * fallback --> hdfs://localhost:xxx/HDFSUser/
+   *
+   * It will find fallback link, but root is not accessible and read only.
+   */
+  @Test(expected = AccessControlException.class, timeout = 30000)
+  public void testCreateOnRootShouldFailEvenFallBackMountLinkConfigured()
+      throws Exception {
+    final Path hdfsTargetPath = new Path(defaultFSURI + HDFS_USER_FOLDER);
+    final Path localTragetPath = new Path(localTargetDir.toURI());
+
+    createLinks(true, hdfsTargetPath, localTragetPath);
+
+    ViewFsOverloadScheme fs = (ViewFsOverloadScheme) FileSystem.get(conf);
+    try {
+      fs.create(new Path("/onRootWhenFallBack"));
+      Assert.fail(
+          "It should fail as root is read only in viewFS, even when configured"
+              + " with fallback.");
+    } finally {
+      fs.close();
+    }
+  }
+
+  /**
+   * Create mount links as follows
+   * hdfs://localhost:xxx/HDFSUser --> hdfs://localhost:xxx/HDFSUser/
+   * hdfs://localhost:xxx/local --> file://TEST_ROOT_DIR/root/
+   * fallback --> hdfs://localhost:xxx/HDFSUser/
+   *
+   * It will find fallback link, but root is not accessible and read only.
+   */
+  @Test(expected = UnsupportedFileSystemException.class, timeout = 30000)
+  public void testInvalidOverloadSchemeTargetFS() throws Exception {
+    final Path hdfsTargetPath = new Path(defaultFSURI + HDFS_USER_FOLDER);
+    final Path localTragetPath = new Path(localTargetDir.toURI());
+    conf = new Configuration();
+    createLinks(true, hdfsTargetPath, localTragetPath);
```

Comment:
calling createLinks isn't even needed here right?

Reply:
Without creating links ViewFS fails with "not mount points configured". So, to move forward to the intended scenario, we just need to create links.